patches/server/0694-Add-more-advancement-API.patch

```diff
@@ -73,6 +80,47 @@ index 0000000000000000000000000000000000000000..0567e500c40d3d424ddc19062c4f6da9
 +        };
 +    }
 +}
+diff --git a/src/main/java/net/minecraft/advancements/Advancement.java b/src/main/java/net/minecraft/advancements/Advancement.java
+index 2c01231dcfbb992c9d2f034fcfd4af52fe8ac265..9219ab40bb3a88d7fad80287d159be500f96f2f3 100644
+--- a/src/main/java/net/minecraft/advancements/Advancement.java
++++ b/src/main/java/net/minecraft/advancements/Advancement.java
+@@ -53,20 +53,27 @@ public class Advancement {
+             parent.addChild(this);
+         }
+ 
+-        if (display == null) {
+-            this.chatComponent = Component.literal(id.toString());
++        this.chatComponent = constructDisplayComponent(this.id, this.display); // Paper - moved to static method
++
++    }
++    // Paper start - copied from constructor
++    public static Component constructDisplayComponent(final @Nullable ResourceLocation id, final @Nullable DisplayInfo display) {
++        if (id == null && display == null) {
++            throw new IllegalArgumentException("can't both be null");
++        } else if (display == null) {
++            return Component.literal(id.toString());
+         } else {
+-            Component ichatbasecomponent = display.getTitle();
+-            ChatFormatting enumchatformat = display.getFrame().getChatColor();
+-            MutableComponent ichatmutablecomponent = ComponentUtils.mergeStyles(ichatbasecomponent.copy(), Style.EMPTY.withColor(enumchatformat)).append("\n").append(display.getDescription());
+-            MutableComponent ichatmutablecomponent1 = ichatbasecomponent.copy().withStyle((chatmodifier) -> {
+-                return chatmodifier.withHoverEvent(new HoverEvent(HoverEvent.Action.SHOW_TEXT, ichatmutablecomponent));
++            Component title = display.getTitle();
```

Comment:
this seems like more diff than necessary?

Reply:
yeah, that turned out way worse than I had hoped for. I got it better tho.